### PR TITLE
Resolve markdown links to external packages

### LINKED ISSUE #1612: Always resolve to "anchored" Rd xrefs
In order to improve cross-package links in the new single-page-html documentation, CRAN will start showing NOTEs for Rd pages which link to a topic that exists in multiple CRAN packages, such as `toJSON` (in jsonlite, RJSONIO, rjson).

Seemingly it is not trivial at CMD check time to resolve a topic to a given based on the package imports/suggests. They suggest you use "anchored" Rd links, which simply means using the full `\link[pkg:topic]{topic}` syntax.

If possible, way to reduce some of the fallout would be to resolve a `[topic]` name to the full `\link[pkg:topic]{topic}` at roxygenize time, by inspecting at the package namespace, i.e. to automatically determine the `pkg:` value when unset:

https://github.com/r-lib/roxygen2/blob/2357de6fe50df352f286a4ec77f65e00e582fa21/R/markdown-link.R#L12-L23

### PR DESCRIPTION
**Pull request by gaborcsardi**
- [x] ~~Anchor links to base packages as well.~~
- [x] [Do **not** anchor base packages.](https://github.com/r-lib/roxygen2/issues/1612#issuecomment-2218740209)
- [x] Tests.
- [x] News.
- [x] Docs.
- [x] Cache dependency info.
- [x] Resolve re-exported functions.

Closes #1612.

### CODE REVIEW

File: R/markdown-link.R [hunk 1]
```diff
@@ -11,18 +11,27 @@
 #' ```
 #' MARKDOWN           LINK TEXT  CODE RD
 #' --------           ---------  ---- --
-#' [fun()]            fun()       T   \\link[=fun]{fun()}
-#' [obj]              obj         F   \\link{obj}
+#' [fun()]            fun()       T   \\link[=fun]{fun()} or
+#'                                    \\link[pkg::file]{fun()}
```
Comment (krlmlr) [inline R/markdown-link.R:15]:
```suggestion
#'                                    \\link[pkg:file]{fun()}
```

Similarly elsewhere?

File: R/markdown-link.R [hunk 2]
```diff
@@ -174,6 +185,47 @@ parse_link <- function(destination, contents, state) {
...
+  if (me == "") return(NA_character_)
+
+  # if it is in the current package, then no need for package name, right?
+  if (has_topic(topic, me)) return(NA_character_)
+
+  # try packages in depends, imports, suggests first, error on name clashes
+  pkgdir <- pkgdir %||% roxy_meta_get("current_package_dir")
+  deps <- desc::desc_get_deps(pkgdir)
+  deps <- deps[deps$package != "R", ]
+  deps <- deps[deps$type %in% c("Depeneds", "Imports", "Suggests"), ]
```
Comment (krlmlr) [inline R/markdown-link.R:200]:
```suggestion
  deps <- deps[deps$type %in% c("Depends", "Imports", "Suggests"), ]
```

**Review by krlmlr — COMMENTED** [review]
Works for me in duckdb, thank you for the very quick fix!

File: R/markdown-link.R [hunk 3]
```diff
@@ -11,18 +11,27 @@
 #' ```
 #' MARKDOWN           LINK TEXT  CODE RD
 #' --------           ---------  ---- --
-#' [fun()]            fun()       T   \\link[=fun]{fun()}
-#' [obj]              obj         F   \\link{obj}
+#' [fun()]            fun()       T   \\link[=fun]{fun()} or
+#'                                    \\link[pkg:file]{pkg::fun()}
+#' [obj]              obj         F   \\link{obj} or
+#'                                    \\link[pkg:obj]{pkg::obj}
```
Comment (krlmlr) [inline R/markdown-link.R:17]: Not sure about this, but consistent with `[pkg::obj]` below:

```suggestion
#'                                    \\link[pkg:file]{pkg::obj}
```

File: R/markdown-link.R [hunk 4]
```diff
@@ -174,6 +185,59 @@ parse_link <- function(destination, contents, state) {
   }
 }
 
+resolve_link_package <- function(topic, me = NULL, pkgdir = NULL) {
+  me <- me %||% roxy_meta_get("current_package")
```
Comment (krlmlr) [inline R/markdown-link.R:189]: Can `me` still be `NULL` here? I saw this when calling this function manually.

**Review by krlmlr — COMMENTED** [review]
Super-minor stuff.

Next level: automatically add these qualifications with `@inheritParams` -- I have this in DBI, and I see how niche this probably is.

**Comment by gaborcsardi** [comment]
@krlmlr 
> Next level: automatically add these qualifications with @inheritParams -- I have this in DBI, and I see how niche this probably is.

You mean adjusting the link target packages in Rd from `@inheritParams`? Doesn't that work already?

File: NEWS.md [hunk 5]
```diff
@@ -20,7 +23,7 @@
   (#1563, @krlmlr).
 
 * `@importFrom` works again for quoted non-syntactic names, e.g.
-  `@importFrom magrittr "%>%"` or ``@importFrom rlang `:=` `` 
+  `@importFrom magrittr "%>%"` or ``@importFrom rlang `:=` ``
```
Comment (hadley) [inline NEWS.md:26]: Looks like you accidentally stripped trailing whitespace.

File: R/markdown-link.R [hunk 6]
```diff
@@ -11,18 +11,27 @@
...
+#'                                    \\link[pkg:file]{text}
 #' [text][pkg::fun()] text        F   \\link[pkg:file]{text}
 #' [text][pkg::obj]   text        F   \\link[pkg:file]{text}
-#' [s4-class]         s4          F   \\linkS4class{s4}
+#' [s4-class]         s4          F   \\linkS4class{s4} or
+#'                                    \\link[pkg:file]{s4}
 #' [pkg::s4-class]    pkg::s4     F   \\link[pkg:file]{pkg::s4}
 #' ```
 #'
+#' For the ones that have two RD variants the first version is used for
```
Comment (hadley) [inline R/markdown-link.R:31]:
```suggestion
#' For the links with have two `.Rd` variants the first version is used for
```

File: R/markdown.R [hunk 7]
```diff
@@ -68,6 +70,7 @@ markdown <- function(text, tag = NULL, sections = FALSE) {
 #' @keywords internal
 
 markdown_pass1 <- function(text) {
+  rm(list = ls(envir = mddata), envir = mddata)
```
Comment (hadley) [inline R/markdown.R:73]: Don't we want to cache deps across all calls? In which case, `roxy_meta_get()` might be a better home for this cache?

**Comment by krlmlr** [comment]
It changes things in tibble, trying in CI/CD now: https://github.com/tidyverse/tibble/actions/runs/9871695446/job/27260156746

Thank you very much!

File: R/markdown-link.R [hunk 8]
```diff
@@ -174,6 +185,100 @@ parse_link <- function(destination, contents, state) {
...
+  me <- me %||% roxy_meta_get("current_package")
+  # this is  from the roxygen2 tests, should not happen on a real package
+  if (is.null(me) || is.na(me) || me == "") return(NA_character_)
+
+  # if it is in the current package, then no need for package name, right?
+  if (has_topic(topic, me)) return(NA_character_)
+
+  # try packages in depends, imports, suggests first, error on name clashes
+  pkgdir <- pkgdir %||% roxy_meta_get("current_package_dir")
+  deps <- mddata[["deps"]] <- mddata[["deps"]] %||% desc::desc_get_deps(pkgdir)
```
Comment (hadley) [inline R/markdown-link.R:198]: I found `mddata[["deps"]] <- mddata[["deps"]]` a bit hard to parse mentally.

File: R/markdown-link.R [hunk 9]
```diff
@@ -174,6 +185,100 @@ parse_link <- function(destination, contents, state) {
...
+  deps <- deps[deps$package != "R", ]
+  deps <- deps[deps$type %in% c("Depends", "Imports", "Suggests"), ]
+  pkgs <- deps$package
```
Comment (hadley) [inline R/markdown-link.R:201]: Maybe we only need to cache `pkgs`?

File: R/markdown-link.R [hunk 10]
```diff
@@ -174,6 +185,100 @@ parse_link <- function(destination, contents, state) {
...
+
+  pkg_has_topic <- pkgs[map_lgl(pkgs, has_topic, topic = topic)]
+  pkg_has_topic <- map_chr(pkg_has_topic, function(p) {
+    p0 <- find_reexport_source(topic, p)
+    if (is.na(p0)) p else p0
+  })
+  pkg_has_topic <- unique(pkg_has_topic)
+  base <- base_packages()
+  if (length(pkg_has_topic) == 1) {
```
Comment (hadley) [inline R/markdown-link.R:210]: Maybe explicitly handle `length(pkg_has_topic) == 0`  just to be clear?

File: R/markdown-link.R [hunk 11]
```diff
@@ -174,6 +185,100 @@ parse_link <- function(destination, contents, state) {
...
+
+  # try base packages as well, take the first hit,
+  # there should not be any name clashes, anyway
+  for (bp in base) {
+    if (has_topic(topic, bp)) return(NA_character_)
+  }
+
+  cli::cli_abort(c(
+    "Could not resolve link to topic {.val {topic}} in the dependent
+     and base packages.",
```
Comment (hadley) [inline R/markdown-link.R:231]:
```suggestion
    "Could not resolve link to topic {.val {topic}} in the dependencies or base packages.",
```

**Review by hadley — APPROVED** [review]

**Comment by krlmlr** [comment]
Works for tibble, green builds again!

File: NEWS.md (hunk 5 above)
Comment (gaborcsardi) [reply NEWS.md:26]: Thanks Positron! :/

File: R/markdown.R (hunk 7 above)
Comment (gaborcsardi) [reply R/markdown.R:73]: It is fairly fast to do it once, so it is safer not to cache it across all calls, in case the dependencies change.

File: R/markdown-link.R (hunk 8 above)
Comment (gaborcsardi) [reply R/markdown-link.R:198]: Yeah. But it is so much simpler than an `if () { } else {}`. Anyway, I can change it to an `if`.

**Comment by gaborcsardi** [comment]
So, I have been using this for a bit, and it does not work very well when you are adding new functions that are not documented yet, or refactoring a package and renaming functions. The errors don't allow the changes to propagate properly.

One solution would be to have a warning instead of an error. This is easy to implement. The caveat is that you might end up with incorrect links, unless you run `document()` twice.

Another solution is to delay the resolution of the link target packages until the (new) topic of the documented package are known. This might be difficult to implement with how roxygen2 works currently. We would essentially need to do the final link fixup in the generated Rd docs.

**Comment by krlmlr** [comment]
To better understand this: the problem is that this PR works worse than the main branch if documentation is added to a function that didn't have documentation before? What errors are you seeing?

**Comment by hadley** [comment]
I think switching to warnings are probably a good start. Can we assume that if we can't find the topic it's in the current package?

**Comment by gaborcsardi** [comment]
Good idea. R CMD check will catch the real errors anyway.

**Comment by gaborcsardi** [comment]
@krlmlr Yeah, when you are refactoring or developing, you might have a state where you are referring to a function that is not documented. E.g. if you simply rename a function that you are cross-linking from a manual page. With this PR the package of the renamed function cannot be resolved, which causes an error, and the manual pages are not updated. So you are stuck in this state.

**Comment by gaborcsardi** [comment]
@hadley OK, I turned them into warnings. Can you please take another look?

File: R/markdown-link.R [hunk 12]
```diff
@@ -174,6 +185,100 @@ parse_link <- function(destination, contents, state) {
...
+  if (length(pkg_has_topic) == 0) {
+    # fall through to check base packages as well
+  } else if (length(pkg_has_topic) == 1) {
+    if (pkg_has_topic %in% base) {
+      return(NA_character_)
+    } else {
+      return(pkg_has_topic)
+    }
+  } else {
+    cli::cli_warn(c(
```
Comment (hadley) [inline R/markdown-link.R:215]: Can you use `warn_roxy_topic()` here (and below)?

File: R/markdown-link.R [hunk 13]
```diff
@@ -174,6 +185,100 @@ parse_link <- function(destination, contents, state) {
...
+  # try base packages as well, take the first hit,
+  # there should not be any name clashes, anyway
+  for (bp in base) {
+    if (has_topic(topic, bp)) return(NA_character_)
+  }
+
+  cli::cli_warn(c(
+    "Could not resolve link to topic {.val {topic}} in the dependencies or base packages.",
+    "i" = paste(
+      "If you haven'r documented {.val {topic}} yet, or just changed its name, this is normal.",
```
Comment (hadley) [inline R/markdown-link.R:230]: I had a vague memory that roxygen2 maintains some index of topics for the current package, but I can't find any sign of it in the code, so I'm confident that this is the best you can do

**Review by hadley — APPROVED** [review]

File: R/markdown-link.R (hunk 12 above)
Comment (gaborcsardi) [reply R/markdown-link.R:215]: I don't have the topic here I think. I am not even sure that we have a topic at this point. Maybe I could use `warn_roxy_tag()` if we pass the tag through the machinery?

Comment (hadley) [reply R/markdown-link.R:215]: Oh I pattern matched on the wrong `topic`.  I think it would be worth passing all the necessary info down to so `warn_roxy_tag()` can give a file name/line number.

File: R/markdown-link.R (hunk 13 above)
Comment (gaborcsardi) [reply R/markdown-link.R:230]: Yeah, we don't have the topics yet here. This was basically my second idea above, i.e. wait until we have the topics and do the substitution after that. Probably possible, but I think the warnings are also ok.

File: R/markdown-link.R (hunk 12 above)
Comment (gaborcsardi) [reply R/markdown-link.R]: This is was pretty easy in the end.

**Comment by jeroen** [comment]
Great, this worked for all cases I tried!

**Review by jeroen — APPROVED** [review]
